pkg/controller/risingwave_compute_sts_update_strategy.go

```diff
@@ -0,0 +1,166 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	risingwavev1alpha1 "github.com/risingwavelabs/risingwave-operator/apis/risingwave/v1alpha1"
+	"github.com/risingwavelabs/risingwave-operator/pkg/consts"
+	"github.com/risingwavelabs/risingwave-operator/pkg/metrics"
+	"github.com/risingwavelabs/risingwave-operator/pkg/utils"
+)
+
+// RisingWaveComputeSTSUpdateStrategy implements an optimized update strategy for the StatefulSets that controls the compute nodes.
+type RisingWaveComputeSTSUpdateStrategy struct {
+	client client.Client
+}
+
+// Reconcile implements reconcile.Reconciler.
+func (s *RisingWaveComputeSTSUpdateStrategy) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// StatefulSet that was updated.
+	var sts appsv1.StatefulSet
+
+	err := s.client.Get(ctx, request.NamespacedName, &sts)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(10).Info("Not found, abort")
+			return ctrlkit.NoRequeue()
+		}
+		logger.Error(err, "Failed to get StatefulSet")
+		return ctrlkit.RequeueIfErrorAndWrap("unable to get statefulset", err)
+	}
+
+	// Controller manager hasn't observed the change. Wait for another second and retry.
+	if sts.Generation != sts.Status.ObservedGeneration {
+		return ctrlkit.RequeueAfter(time.Second)
+	}
+
+	// Return immediately if the StatefulSet isn't updating.
+	if sts.Status.UpdatedReplicas == sts.Status.Replicas {
+		return ctrlkit.NoRequeue()
+	}
+
+	labelSelector, err := metav1.LabelSelectorAsSelector(sts.Spec.Selector)
+	if err != nil {
+		logger.Error(err, "Failed to get label selector")
+		return ctrlkit.RequeueIfErrorAndWrap("unable to get label selector", err)
+	}
+
+	// Pods of the StatefulSet.
+	var podList corev1.PodList
+	err = s.client.List(ctx, &podList, client.InNamespace(request.Namespace), client.MatchingLabelsSelector{Selector: labelSelector})
+	if err != nil {
+		logger.Error(err, "Failed to list Pods")
+		return ctrlkit.RequeueIfErrorAndWrap("unable to list pods", err)
+	}
+
+	// A pod is considered up-to-date if and only if its revision is greater than or equal to the update revision.
+	isPodUpToDate := func(pod *corev1.Pod) bool {
+		controllerRevisionHash := pod.GetLabels()[appsv1.StatefulSetRevisionLabel]
+
+		return controllerRevisionHash == sts.Status.UpdateRevision
+	}
+
+	// Delete pods that is outdated immediately.
+	toDeletePods := lo.Filter(podList.Items, func(pod corev1.Pod, _ int) bool {
+		return pod.DeletionTimestamp.IsZero() &&
+			metav1.GetControllerOfNoCopy(&pod).UID == sts.UID &&
+			!isPodUpToDate(&pod)
+	})
+	if len(toDeletePods) == 0 {
+		return ctrlkit.NoRequeue()
+	}
+
+	logger.V(1).Info("Deleting pods...", "pods", utils.MapObjectsToNames[corev1.Pod, *corev1.Pod](toDeletePods))
+	for _, pod := range toDeletePods {
+		err := s.client.Delete(ctx, &pod, client.Preconditions{UID: &pod.UID})
```

Comment:
Will all the missing pods be started concurrently after deletion?

Reply:
Exactly. It is determined by the [pod management policy](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#pod-management-policies), which is always `Parallel` in our case.